### PR TITLE
B2CA-1558: Port Standard Makefile

### PR DESCRIPTION
**Pull request by cedelavergne-ledger**
This PR cleans up the Makefile, and few fixes in the code:

- Port current config on Standard App Makefile
- Remove useless clones
  - `flare`
  - `flare_coston`
  - `ropsten`
  -  `goerli`
- Update workflows
  - Add _recursive_ checkout instead of forcing `git submodule` from the `Makefile` (and adapt the documentation)
  - Remove useless workflow for `ethereum-plugin-sdk` auto commit
- Remove useless screenshots for `test_legacy_contract`, and simplify associated test
- Code improvements / Fix warnings
  - CodeQL
  - Report most modifications from PR #225 
  - replace `array_hexstr` and `%*H` format by sdk function `format_hex`
  - Add `noreturn` attribute in `main.c`
- Cleanup plugin-sdk: No need to build or copy files locally, or keep local copy of `etherum-plugin-sdk` files
- Remove `U2F` support

## Changes include

- [X] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [X] Tests
- [X] Documentation
- [X] Other (for changes that might not fit in any category)

### CODE REVIEW

File: tests/unit/CMakeLists.txt [hunk 1]
```diff
@@ -52,7 +52,7 @@ FetchContent_MakeAvailable(cmocka)
 add_compile_definitions(TEST DEBUG=0 SKIP_FOR_CMOCKA)
 
 include_directories(../../src/)
-include_directories(../../src_common)
+include_directories(../../src_plugin_sdk)
```
Comment (apaillier-ledger) [inline tests/unit/CMakeLists.txt:55]: This directory does not exist.

**Review by apaillier-ledger — CHANGES_REQUESTED** [review]

File: Makefile [hunk 2]
```diff
@@ -315,11 +65,102 @@ $(NETWORK_ICONS_FILE):
...
+CURVE_APP_LOAD_PARAMS += secp256k1
+
+# Application allowed derivation paths.
+# You should request a specific path for your app.
+# This serve as an isolation mechanism.
+# Most application will have to request a path according to the BIP-0044
+# and SLIP-0044 standards.
+# If your app needs it, you can specify multiple path by using:
+# `PATH_APP_LOAD_PARAMS = "44'/1'" "45'/1'"`
+PATH_APP_LOAD_PARAMS += "45'" "1517992542'/1101353413'" "44'/1'"
```
[the remark this answers is not in the document]
Comment (cedelavergne-ledger) [reply Makefile:96]: Done

File: tests/unit/CMakeLists.txt (hunk 1 above)
Comment (cedelavergne-ledger) [reply tests/unit/CMakeLists.txt:55]: Fixed

File: Makefile [hunk 3]
```diff
@@ -21,289 +21,39 @@ endif
...
-#########################
-# Set Chain environment #
-#########################
+########################################
+#        Mandatory configuration       #
+########################################
 
 ifeq ($(CHAIN),)
     CHAIN = ethereum
+    APPNAME = Ethereum
```
Comment (apaillier-ledger) [inline Makefile:30]: This should not be needed since it is already declared in `makefile_conf/chain/ethereum.mk`

File: Makefile [hunk 4]
```diff
@@ -21,289 +21,39 @@ endif
...
+ifeq ($(TARGET_NAME),TARGET_STAX)
     APP_SOURCE_PATH += src_nbgl
 else
-    SDK_SOURCE_PATH += lib_ux
     APP_SOURCE_PATH += src_bagl
 endif
-
-# Allow usage of function from lib_standard_app/crypto_helpers.c
+APP_SOURCE_FILES += ./ethereum-plugin-sdk/src/common_utils.c
+APP_SOURCE_FILES += ./ethereum-plugin-sdk/src/plugin_utils.c
```
Comment (fbeutin-ledger) [inline Makefile:52]: `APP_SOURCE_FILES += $(filter-out ./ethereum-plugin-sdk/src/main.c, $(wildcard ./ethereum-plugin-sdk/src/*.c))`

File: Makefile (hunk 3 above)
Comment (cedelavergne-ledger) [reply Makefile:30]: Yes, but a reason I didn't explain, using the VSCode extension, I cannot delete the App from the device if the name here is `ethereum`, whereas it works with `Ethereum`.

File: Makefile (hunk 4 above)
Comment (cedelavergne-ledger) [reply Makefile:52]: Good idea. I have used your proposal

File: Makefile [hunk 5]
```diff
@@ -315,11 +65,102 @@ $(NETWORK_ICONS_FILE):
...
+#prepare hsm generation
+ifeq ($(TARGET_NAME),TARGET_STAX)
+    DEFINES += ICONGLYPH=C_stax_chain_$(CHAIN_ID)_64px
+    DEFINES += ICONBITMAP=C_stax_chain_$(CHAIN_ID)_64px_bitmap
+    DEFINES += ICONGLYPH_SMALL=C_stax_chain_$(CHAIN_ID)
+endif
+
+
+# Application allowed derivation curves.
+# Possibles curves are: secp256k1, secp256r1, ed25519 and bls12381g1
```
Comment (fbeutin-ledger) [inline Makefile:83]: cleanup

File: Makefile (hunk 3 above)
Comment (apaillier-ledger) [reply Makefile:30]: Then it's a VS Code extension problem.

File: Makefile [hunk 6]
```diff
@@ -315,11 +64,102 @@ $(NETWORK_ICONS_FILE):
...
+# If your app needs it, you can specify multiple curves by using:
+# `CURVE_APP_LOAD_PARAMS = <curve1> <curve2>`
+CURVE_APP_LOAD_PARAMS += secp256k1
+
+# Application allowed derivation paths.
+# You should request a specific path for your app.
```
Comment (fbeutin-ledger) [inline Makefile:89]: cleanup

File: Makefile [hunk 7]
```diff
@@ -315,11 +64,102 @@ $(NETWORK_ICONS_FILE):
...
+# This serve as an isolation mechanism.
+# Most application will have to request a path according to the BIP-0044
+# and SLIP-0044 standards.
+# If your app needs it, you can specify multiple path by using:
+# `PATH_APP_LOAD_PARAMS = "44'/1'" "45'/1'"`
+PATH_APP_LOAD_PARAMS += "45'" "44'/1'"
+
+# Setting to allow building variant applications
+# - <VARIANT_PARAM> is the name of the parameter which should be set
+#   to specify the variant that should be build.
```
Comment (fbeutin-ledger) [inline Makefile:99]: cleanup

File: Makefile [hunk 8]
```diff
@@ -315,11 +64,102 @@ $(NETWORK_ICONS_FILE):
...
+# - <VARIANT_VALUES> a list of variant that can be build using this app code.
+#   * It must at least contains one value.
+#   * Values can be the app ticker or anything else but should be unique.
+VARIANT_PARAM = CHAIN
+VARIANT_VALUES = $(SUPPORTED_CHAINS)
+
+# Activate dependency only for specific CHAIN
+ifneq ($(CHAIN),ethereum)
+DEP_APP_LOAD_PARAMS = Ethereum:$(APPVERSION)
```
Comment (fbeutin-ledger) [inline Makefile:108]: indent

File: Makefile [hunk 9]
```diff
@@ -315,11 +64,102 @@ $(NETWORK_ICONS_FILE):
...
+DEFINES_LIB = USE_LIB_ETHEREUM
+endif
+
+# Enabling DEBUG flag will enable PRINTF and disable optimizations
+#DEBUG = 1
```
Comment (fbeutin-ledger) [inline Makefile:113]: can be removed, we use the command line

File: makefile_conf/chain/ethereum.mk [hunk 10]
```diff
@@ -1,11 +1,10 @@
...
 CHAIN_ID = 1
 # Allow to derive ETH 2 public keys
-APP_LOAD_PARAMS += --path "12381/3600" --curve bls12381g1
+PATH_APP_LOAD_PARAMS += "12381/3600"
+CURVE_APP_LOAD_PARAMS += bls12381g1
 DEFINES += HAVE_ETH2
 APPNAME = "Ethereum"
-DEFINES_LIB=
 DEFINES += HAVE_BOLOS_APP_STACK_CANARY
-APP_LOAD_FLAGS=--appFlags 0xa40
```
Comment (fbeutin-ledger) [inline makefile_conf/chain/ethereum.mk:11]: Il est bien set plus tard ?

File: makefile_conf/features.mk [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+# ENS
+ifneq ($(TARGET_NAME),TARGET_NANOS)
+    DEFINES += HAVE_DOMAIN_NAME
+    DOMAIN_NAME_TEST_KEY ?= 0
+    ifneq ($(DOMAIN_NAME_TEST_KEY),0)
+        DEFINES += HAVE_DOMAIN_NAME_TEST_KEY
+    endif
+endif
+
+ifneq ($(NOCONSENT),)
```
Comment (fbeutin-ledger) [inline makefile_conf/features.mk:66]: Use same pattern
```
NOCONSENT =? 0
ifneq ($(NOCONSENT),0)
```

There is a point in removing this flag entirely though ? @apaillier-ledger 
That's for another PR, but I don't see the benefit of this flag now that we have good testing support and test keys

File: Makefile (hunk 3 above)
Comment (cedelavergne-ledger) [reply Makefile:30]: Adding a comment to remember removing the APPNAME definition here, when VSCode extension will be updated

File: Makefile (hunk 8 above)
Comment (cedelavergne-ledger) [reply Makefile:108]: Done

File: src/uint128.c [hunk 12]
```diff
@@ -53,11 +53,9 @@ void shiftl128(const uint128_t *const number, uint32_t value, uint128_t *const t
     } else if (value < 64) {
         UPPER_P(target) = (UPPER_P(number) << value) + (LOWER_P(number) >> (64 - value));
         LOWER_P(target) = (LOWER_P(number) << value);
-    } else if ((128 > value) && (value > 64)) {
+    } else {
         UPPER_P(target) = LOWER_P(number) << (value - 64);
```
Comment (fbeutin-ledger) [inline src/uint128.c:57]: Add a comment to remember that we are in the case `128 > value) && (value > 64`

File: src/uint128.c [hunk 13]
```diff
@@ -74,11 +72,9 @@ void shiftr128(const uint128_t *const number, uint32_t value, uint128_t *const t
         UPPER(result) = UPPER_P(number) >> value;
         LOWER(result) = (UPPER_P(number) << (64 - value)) + (LOWER_P(number) >> value);
         copy128(target, &result);
-    } else if ((128 > value) && (value > 64)) {
+    } else {
         LOWER_P(target) = UPPER_P(number) >> (value - 64);
```
Comment (fbeutin-ledger) [inline src/uint128.c:76]: same

File: src/uint_common.c [hunk 14]
```diff
@@ -58,3 +58,16 @@ void reverseString(char *const str, uint32_t length) {
         str[j] = c;
     }
 }
+
+int bytes_to_string(char *out, size_t outl, const void *value, size_t len) {
+    if (strlcpy(out, "0x", outl) != 2) {
+        goto err;
+    }
+    if (format_hex(value, len, out + 2, outl - 2) == -1) {
```
Comment (fbeutin-ledger) [inline src/uint_common.c:66]: `< 0` seem more standard

File: src/uint_common.c [hunk 15]
```diff
@@ -58,3 +58,16 @@ void reverseString(char *const str, uint32_t length) {
...
+        goto err;
+    }
+    return 0;
+err:
+    *out = '\0';
```
Comment (fbeutin-ledger) [inline src/uint_common.c:71]:
```
if (outl >= 1) {
    out[0] = '\0';
}
```
Probably useless but might prevent a nasty error if we call with outl==0

File: src/uint_common.h [hunk 16]
```diff
@@ -32,4 +35,6 @@ void read_u64_be(const uint8_t *const in, uint64_t *const out);
 uint64_t readUint64BE(const uint8_t *const buffer);
 void reverseString(char *const str, uint32_t length);
 
+int bytes_to_string(char *out, size_t outl, const void *value, size_t len);
```
Comment (fbeutin-ledger) [inline src/uint_common.h:38]: Could be nice to move this to the plugin sdk in a second time

File: src_features/performPrivacyOperation/feature_performPrivacyOperation.h [hunk 17]
```diff
@@ -1,3 +1,8 @@
+#ifndef _PERFORM_PRIVACY_OPERATION_H_
```
Comment (fbeutin-ledger) [inline src_features/performPrivacyOperation/feature_performPrivacyOperation.h:1]: I prefer the `#pragma once` pattern but as you want

File: src/uint_common.c (hunk 15 above)
Comment (cedelavergne-ledger) [reply src/uint_common.c:71]: Better to check `outl` at the begining of the function, to ensure the possibility, at least to write `0x`.

File: src_features/performPrivacyOperation/feature_performPrivacyOperation.h (hunk 17 above)
Comment (cedelavergne-ledger) [reply src_features/performPrivacyOperation/feature_performPrivacyOperation.h:1]: Just to be consistent with other headers in `src_features/*`...

File: src/uint_common.h (hunk 16 above)
Comment (cedelavergne-ledger) [reply src/uint_common.h:38]: Agree

File: src/uint_common.c (hunk 15 above)
Comment (fbeutin-ledger) [reply src/uint_common.c:71]: Fine by me, one or the other

File: makefile_conf/features.mk (hunk 11 above)
Comment (apaillier-ledger) [reply makefile_conf/features.mk:66]: Indeed we should remove it in another PR, there is no use for it.

**Pull request merged**